enhancements/security/x509-trust.md

```diff
@@ -0,0 +1,183 @@
+Cluster components may also perform other X.509 verification besides TLS requests.
+For example, user mail processing components may perform [S/MIME][smime] verification.
+This enhancement provides a framework for managing X.509 trust for these situations.
+
+Currently the default trust for CRI-O, the kubelet's cloud provider, and other node-namespace egress is the node's X.509 certificate authority (CA) store, which depends on the operating system ([Red Hat Enterprise Linux CoreOS (RHCOS)][rhcos], etc.) and any subsequent configuration ([the machine-config operator][machine-config-operator], etc.).
+The default trust for containers like the machine-API operator is that container's CA store, which depends on the base layer ([Red Hat Universal Base Image (UBI)][ubi], etc.).
+Many external components will have X.509 certificates signed by one of those default CAs, but this is not always the case.
+
+This enhancement provides a mechanism for configuring additional or alternative trust bundles for all cluster services when communicating with external components whose X.509 certificates are not signed by a default CA.
+
+## Motivation
+
+### Goals
+
+Administrators will have a single location to configure a default additional trust bundle to be used by all cluster components communicating with external components.
+They will also be able to specify whether this default is to be use in addition to, or instead of, the component's default trust bundle.
+
+Administrators will have per-component configuration locations to adjust the cluster-scoped default.
+As for the cluster-scoped default, they will be able to specify whether the per-component trust bundle is to be used in addition to, or instead of, the cluster-scoped default.
+
+### Non-Goals
+
+This enhancement does not address components making intra-cluster TLS requests.
+Trust for those requests is covered by [the Kubernetes API-server Certificates enhancement](../kube-apiserver/certificates) and [the Automated Service CA Rotation enhancement](../automated-service-ca-rotation.md).
+
+This enhancement does not address configuration for non-CA TLS settings like ciphers and allowed TLS versions.
+[The ingress TLS configuration enhancement](../kube-apiserver/tls-config.md) covers those settings for cluster-hosted services.
+Outgoing TLS probably needs similar configuration, but that configuration is decoupled from X.509 trust and should be addressed in a separate enhancement.
+
+## Proposal
+
+### Cluster-scoped Default Trust
+
+There will be a ConfigMap (FIXME: or should this be a new CRD under openshift/api?) named `default-ca-bundle` in the `openshift-config` namespace.
+The ConfigMap will have the following keys:
+
+* `ca-bundle.crt`, containing a PEM-encoded X.509 certificate bundle.
+* `composition`, configuring how the bundle in `ca-bundle.crt` is composed with the component's default trust store.
```

Comment:
I think we should provide an API that points to configmap which has a default location..
and then composition enums are better tracked with validations..

Reply:
> ... then composition enums are better tracked with validations.

The plan is for the composition enums to percolate down through the whole chain, so you can, for example, set a global default and then append to it in the proxy-specific config.  Hmm, that means my schema isn't quite right, because you'd need knobs for both directions "I'm (clobbering|appending to) the default config and I expect that to (clobber|be appended to) the per-component defaults".  I'll see if I can get that addressed...

Back to versioning, if we want to address that specifically, we should make this a CRD instead of using ConfigMaps.  It's probably a good idea to do that, but it sort of sinks this approach as a source of 4.3 stopgaps.  Thoughts?

Reply:
> The plan is for the composition enums to percolate down through the whole chain...

I've pushed f451e19 -> 7d62833 to try and clarify this.

> Back to versioning, if we want to address that specifically, we should make this a CRD instead of using ConfigMaps...

Thinking this over, this won't work for the final per-cluster ConfigMap stage, because we can't mount custom resources into containers.  If we're comfortable with the composition enums in a ConfigMap there, I don't see why we'd be uncomfortable having ConfigMaps further up the chain.  I've pushed 7d62833 -> 00ead3d to try and clarify this.